cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,269 @@
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.metadata.lineage.field;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.api.lineage.field.EndPoint;
+import co.cask.cdap.api.lineage.field.Operation;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.proto.codec.OperationTypeAdapter;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset to store/retrieve field level lineage information.
+ */
+public class FieldLineageDataset extends AbstractDataset {
+
+  // Storage format
+  // --------------
+  //
+  // Checksum associated with the graph:
+  //                               -------------------------------------------------------
+  //                               |            r          |          p                  |
+  // -------------------------------------------------------------------------------------
+  // | checksum | <checksum-value> | Collection<Operation> | <Presentation layer format> |
+  // -------------------------------------------------------------------------------------
+  //
+  // Run level row:
+  //                                                                           --------------------
+  //                                                                           |         c        |
+  // ----------------------------------------------------------------------------------------------
```

Comment:
`b(backward)` and `f(forward)` are for the lineage directions. For example if the pipeline reads from EndPoint `source` and writes to  `destination`, then the operations will provide lineage in backward direction from the endpoint `destination` perspective. From source perspective these will be lineage in forward direction. So we add 2 rows, one for `destination` with `b` marker and another for `source` with `f` marker.

Reply:
I am not sure if we have a use case of retrieving field level lineage per run level. If we do, then it will need to be stored as separate row with key starting with run id. Let me know if this is needed.